src/main/kotlin/com/safehill/kclient/tasks/BackgroundTaskProcessor.kt

```diff
@@ -0,0 +1,39 @@
+package com.safehill.kclient.tasks
+
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import java.util.concurrent.ConcurrentLinkedQueue
+
+class BackgroundTaskProcessor<T: BackgroundTask> {
+
+    private val taskQueue = ConcurrentLinkedQueue<T>()
+    internal var isProcessing = false
+
+    fun run(task: T) {
+        taskQueue.add(task)
+        processTasks()
+    }
+
+    private fun processTasks() {
+        if (!isProcessing) {
+            isProcessing = true
+            GlobalScope.launch {
+                while (taskQueue.isNotEmpty()) {
+                    val task = taskQueue.poll()
+                    task?.run()
+                }
+                isProcessing = false
+            }
+        }
+    }
+
+    fun runRepeatedly(task: T, intervalMillis: Long) {
+        GlobalScope.launch {
```

Comment:
We would need a mechanism to cancel the task as well. We would need a coroutineScope(A user bound coroutine scope or a scope whose cancellation can be triggered properly in a desired moment) to be injected into the BackgroundTaskProcessor for handling the lifecycle of the job. 
For now, lets run with GlobalScope and we can refactor it later once the working mechanism is properly implemented.

Reply:
Both great points